### PR TITLE
enh: declare integer variables before visiting nested function in llvm backend

### LINKED ISSUE #4536: bug: assignment to array with non-constant dimension in subroutine fails
### Minimal Reproducible Example

```fortran
PROGRAM main

   IMPLICIT NONE

   INTEGER :: nang = 2

   CALL my_subroutine

CONTAINS

   SUBROUTINE my_subroutine

      INTEGER, DIMENSION(nang) :: fxhv

      fxhv = [1, 2]

      PRINT *, fxhv

   END SUBROUTINE my_subroutine

END PROGRAM main
```

#### GFortran

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ gfortran ./examples/example.f90 && ./a.out
           1           2
```

#### LFortran

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 360, in __libc_start_main_impl()
  File "./csu/../sysdeps/x86/libc-start.c", line 58, in __libc_start_call_main()
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 2481, in main()
    return main_app(argc, argv);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 2442, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 954, in compile_to_object_file()
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/lfortran/fortran_evaluator.cpp", line 363, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10012, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 5134, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 5110, in visit_asr_t<LCompilers::ASRToLLVMVisitor>()
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 5135, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 4842, in visit_unit_t<LCompilers::ASRToLLVMVisitor>()
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 998, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_symbol(*item.second);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 5137, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 4850, in visit_symbol_t<LCompilers::ASRToLLVMVisitor>()
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3166, in LCompilers::ASRToLLVMVisitor::visit_Program(LCompilers::ASR::Program_t const&)
    visit_procedures(x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 4144, in void LCompilers::ASRToLLVMVisitor::visit_procedures<LCompilers::ASR::Program_t>(LCompilers::ASR::Program_t const&)
    visit_Function(*s);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3900, in LCompilers::ASRToLLVMVisitor::visit_Function(LCompilers::ASR::Function_t const&)
    generate_function(x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 4093, in LCompilers::ASRToLLVMVisitor::generate_function(LCompilers::ASR::Function_t const&)
    define_function_entry(x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 4013, in LCompilers::ASRToLLVMVisitor::define_function_entry(LCompilers::ASR::Function_t const&)
    declare_local_vars(x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3872, in void LCompilers::ASRToLLVMVisitor::declare_local_vars<LCompilers::ASR::Function_t>(LCompilers::ASR::Function_t const&)
    declare_vars(x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3610, in void LCompilers::ASRToLLVMVisitor::declare_vars<LCompilers::ASR::Function_t>(LCompilers::ASR::Function_t const&, bool)
    this->visit_expr_wrapper(m_dims[i].m_length, true);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 5397, in LCompilers::ASRToLLVMVisitor::visit_expr_wrapper(LCompilers::ASR::expr_t*, bool)
    this->visit_expr(*x);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 5202, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/../libasr/asr.h", line 4987, in visit_expr_t<LCompilers::ASRToLLVMVisitor>()
    case exprType::Var: { v.visit_Var((const Var_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6948, in LCompilers::ASRToLLVMVisitor::visit_Var(LCompilers::ASR::Var_t const&)
    fetch_var(v);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6937, in LCompilers::ASRToLLVMVisitor::fetch_var(LCompilers::ASR::Variable_t*)
    fetch_val(x);
AssertFailed: llvm_symtab.find(x_h) != llvm_symtab.end()
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4536.

This only works for Program, we need to apply similar diff as in `visit_Program` to `visit_Module`. @kmr-srbh would you be willing to do once this PR is merged? I think this will be a good reference to learn how things work in llvm backend.

### CODE REVIEW

**Comment by kmr-srbh** [comment]
Thanks you @Pranavchiku, I would be glad to work on this!

**Comment by certik** [comment]
I think this is fine. It's mostly just moving things around in the LLVM backend, right? If it requires some complex accounting in the backend, then we need to design ASR differently for this.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -363,6 +363,7 @@ RUN(NAME subroutines_08 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm)
 RUN(NAME subroutines_09 LABELS gfortran)
 RUN(NAME subroutines_10 LABELS gfortran)
 RUN(NAME subroutines_11 LABELS gfortran)
+RUN(NAME subroutines_12 LABELS gfortran llvm)
```
Comment (Pranavchiku) [inline integration_tests/CMakeLists.txt:366]: I removed wasm and cpp labels, I think CI shall pass now.

**Review by certik — APPROVED** [review]

File: tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr [hunk 2]
```diff
@@ -1,9 +1,9 @@
-  File "tests/errors/runtime_stacktrace_01.f90", line 3
-    print *, main()
-  File "tests/errors/runtime_stacktrace_01.f90", line 10
-    print *, g()
-  File "tests/errors/runtime_stacktrace_01.f90", line 20
-    call f()
-  File "tests/errors/runtime_stacktrace_01.f90", line 14
-    stop 1
-STOP 1
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr:9]: This needs to be fixed: https://github.com/lfortran/lfortran/issues/5474

Comment (Pranavchiku) [reply tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr:9]: Yes, thank you for noticing!

**Pull request merged**